### PR TITLE
test: 100% coverage of fy-select-modal component

### PR DESCRIPTION
**Pull request by suyashpatil78**
Clickup - https://app.clickup.com/t/85zrxnafj

Coverage -

Screenshot 2023-05-08 at 10 50 49 PM

### CODE REVIEW

**Review by Julias0 — CHANGES_REQUESTED** [review]
Will review after @yksolanki9 's review

File: src/app/shared/components/fy-select/fy-select-modal/fy-select-modal.component.spec.ts [hunk 1]
```diff
@@ -1,26 +1,239 @@
...
+    ];
+    component.ngAfterViewInit();
+    dummyHtmlInputElement.value = 'eco';
+    dummyHtmlInputElement.dispatchEvent(new Event('keyup'));
+
+    component.filteredOptions$.pipe(take(1)).subscribe((res) => {
+      expect(res).toEqual([
+        { label: 'None', value: null },
+        { label: 'eco', value: 'eco', selected: false },
+        { label: 'ECONOMY', value: { travel_class: 'ECONOMY', vendor: 'asdf' }, selected: true },
```
Comment (yksolanki9) [inline src/app/shared/components/fy-select/fy-select-modal/fy-select-modal.component.spec.ts:104]: Can value be an object?

Comment (yksolanki9) [reply src/app/shared/components/fy-select/fy-select-modal/fy-select-modal.component.spec.ts:104]: While we're at it, please add types for all inputs in this file

File: src/app/shared/components/fy-select/fy-select-modal/fy-select-modal.component.spec.ts [hunk 2]
```diff
@@ -1,26 +1,239 @@
...
+      ]);
+    });
+    tick(1000);
+  }));
+
+  it('ngAfterViewInit(): should update recentrecentlyUsedItems$', fakeAsync(() => {
+    const dummyHtmlInputElement = document.createElement('input');
```
Comment (yksolanki9) [inline src/app/shared/components/fy-select/fy-select-modal/fy-select-modal.component.spec.ts:111]: Use debugElement

File: src/app/shared/components/fy-select/fy-select-modal/fy-select-modal.component.spec.ts [hunk 3]
```diff
@@ -1,26 +1,239 @@
...
+    dummyHtmlInputElement.dispatchEvent(new Event('keyup'));
+
+    component.filteredOptions$.pipe(take(1)).subscribe((res) => {
+      expect(res).toEqual([
+        { label: 'None', value: null },
+        { label: 'eco', value: 'eco', selected: false },
+        { label: 'ECONOMY', value: { travel_class: 'ECONOMY', vendor: 'asdf' }, selected: true },
+      ]);
+    });
+    tick(1000);
```
Comment (yksolanki9) [inline src/app/shared/components/fy-select/fy-select-modal/fy-select-modal.component.spec.ts:107]: This is synchronous, we don't need tick

File: src/app/shared/components/fy-select/fy-select-modal/fy-select-modal.component.spec.ts [hunk 4]
```diff
@@ -1,26 +1,239 @@
...
+    );
+    utilityService.searchArrayStream.and.returnValue(() => of([{ label: 'business', value: 'BUSINESS' }]));
+
+    component.ngAfterViewInit();
+    dummyHtmlInputElement.value = 'busi';
+    dummyHtmlInputElement.dispatchEvent(new Event('keyup'));
+    component.recentrecentlyUsedItems$.pipe(take(1)).subscribe((res) => {
+      expect(res).toEqual([{ label: 'business', value: 'BUSINESS' }]);
+    });
+    tick(1000);
```
Comment (yksolanki9) [inline src/app/shared/components/fy-select/fy-select-modal/fy-select-modal.component.spec.ts:129]: not needed

File: src/app/shared/components/fy-select/fy-select-modal/fy-select-modal.component.spec.ts [hunk 5]
```diff
@@ -1,26 +1,239 @@
...
+  }));
+
+  it('onDoneClick(): should dismiss the modal', () => {
+    component.onDoneClick();
+    expect(modalController.dismiss).toHaveBeenCalledTimes(1);
+  });
+
+  it('clearValue(): should clear the value and trigger a keyup event', () => {
+    const searchBarRefSpy = {
```
Comment (yksolanki9) [inline src/app/shared/components/fy-select/fy-select-modal/fy-select-modal.component.spec.ts:158]: use debugElement

File: src/app/shared/components/fy-select/fy-select-modal/fy-select-modal.component.spec.ts [hunk 6]
```diff
@@ -1,26 +1,239 @@
...
+    component.searchBarRef = searchBarRefSpy;
+    component.clearValue();
+
+    expect(component.value).toBe('');
+    expect(searchBarRefSpy.nativeElement.value).toBe('');
+    expect(searchBarRefSpy.nativeElement.dispatchEvent).toHaveBeenCalledOnceWith(new Event('keyup'));
+  });
+
+  it('should return recently used items from API if available', fakeAsync(() => {
+    component.options = [
```
Comment (yksolanki9) [inline src/app/shared/components/fy-select/fy-select-modal/fy-select-modal.component.spec.ts:171]: We're setting options in every test case, can we set it once in beforeEach and update the specs to use it?

File: src/app/shared/components/fy-select/fy-select-modal/fy-select-modal.component.spec.ts [hunk 7]
```diff
@@ -1,26 +1,239 @@
...
+      { label: 'business', value: 'BUSINESS' },
+      { label: 'economy', value: 'ECONOMY' },
+      { label: 'first class', value: 'FIRST_CLASS' },
+    ];
+    const recentlyUsed = [
```
Comment (yksolanki9) [inline src/app/shared/components/fy-select/fy-select-modal/fy-select-modal.component.spec.ts:176]: Similar for this, if a spec doesn't need it we can overwrite it to be `[]` or undefined for that spec, but initialize data inside beforeEach

File: src/app/shared/components/fy-select/fy-select-modal/fy-select-modal.component.spec.ts [hunk 8]
```diff
@@ -1,26 +1,239 @@
...
+      { label: 'business', value: 'BUSINESS' },
+      { label: 'economy', value: 'ECONOMY' },
+    ];
+
+    component.recentlyUsed = recentlyUsed;
+
+    component.getRecentlyUsedItems().subscribe((result) => {
+      expect(result).toEqual(recentlyUsed);
+    });
+    tick(1000);
```
Comment (yksolanki9) [inline src/app/shared/components/fy-select/fy-select-modal/fy-select-modal.component.spec.ts:186]: not needed

File: src/app/shared/components/fy-select/fy-select-modal/fy-select-modal.component.spec.ts [hunk 9]
```diff
@@ -1,26 +1,239 @@
...
+    ];
+    const filteredItems = [{ label: 'business', value: 'BUSINESS', selected: true }];
+
+    const getSpy = recentLocalStorageItemsService.get.and.returnValue(Promise.resolve(localStorageItems));
+
+    component.recentlyUsed = null;
+    component.currentSelection = 'BUSINESS';
+
+    component.getRecentlyUsedItems().subscribe((result) => {
+      expect(getSpy).toHaveBeenCalledWith(component.cacheName);
```
Comment (yksolanki9) [inline src/app/shared/components/fy-select/fy-select-modal/fy-select-modal.component.spec.ts:207]: toHaveBeenCalledOnceWith

**Review by yksolanki9 — CHANGES_REQUESTED** [review]
Initialize data inside beforeEach instead of defining values for every spec

File: src/app/shared/components/fy-select/fy-select-modal/fy-select-modal.component.spec.ts [hunk 10]
```diff
@@ -1,26 +1,239 @@
...
+    component.nullOption = true;
+    component.customInput = true;
+    component.defaultLabelProp = 'economy';
+    component.options = [
+      { label: 'economy', value: 'ECONOMY' },
+      { label: 'business', value: 'BUSINESS' },
+    ];
+    component.ngAfterViewInit();
+    dummyHtmlInputElement.value = 'econo';
+    dummyHtmlInputElement.dispatchEvent(new Event('keyup'));
```
Comment (yksolanki9) [inline src/app/shared/components/fy-select/fy-select-modal/fy-select-modal.component.spec.ts:71]: Call dispatchEvent twice for 100% coverage

File: src/app/shared/components/fy-select/fy-select-modal/fy-select-modal.component.spec.ts [hunk 11]
```diff
@@ -1,26 +1,239 @@
...
+    utilityService = TestBed.inject(UtilityService) as jasmine.SpyObj<UtilityService>;
+    fixture.detectChanges();
+  }));
 
   it('should create', () => {
     expect(component).toBeTruthy();
   });
+
+  it('ngAfterViewInit(): should update filteredOptions$', fakeAsync(() => {
+    const dummyHtmlInputElement = document.createElement('input');
```
[the remark this answers is not in the document]
Comment (suyashpatil78) [reply src/app/shared/components/fy-select/fy-select-modal/fy-select-modal.component.spec.ts:57]: @yksolanki9 I faced error as `Cannot read property of undefined (reading nativeElement)`.
@Julias0 and I reached upon a solution as in this commit
https://github.com/fylein/fyle-mobile-app/pull/1988/commits/a17d242d2b03e27bcb2d604e3befa1364a5398bf

File: src/app/shared/components/fy-select/fy-select-modal/fy-select-modal.component.spec.ts (hunk 10 above)
Comment (suyashpatil78) [reply src/app/shared/components/fy-select/fy-select-modal/fy-select-modal.component.spec.ts:71]: I have added this on line 128 to get 100% coverage

File: src/app/shared/components/fy-select/fy-select-modal/fy-select-modal.component.spec.ts (hunk 11 above)
[the remark this answers is not in the document]
Comment (yksolanki9) [reply src/app/shared/components/fy-select/fy-select-modal/fy-select-modal.component.spec.ts:57]: https://github.com/fylein/fyle-mobile-app/blob/5659b78707451a89bf6c26b7d164b34ad02f8faa/src/app/shared/components/dependent-fields/dependent-field/dependent-field.component.spec.ts#L103

Check this^. We can discuss it over call if needed

File: src/app/shared/components/fy-select/fy-select-modal/fy-select-modal.component.ts [hunk 12]
```diff
@@ -51,7 +51,7 @@ export class FySelectModalComponent implements OnInit, AfterViewInit {
 
   @Input() label;
 
-  value = '';
+  value: string | object = '';
```
Comment (yksolanki9) [inline src/app/shared/components/fy-select/fy-select-modal/fy-select-modal.component.ts:54]: Write the contract of object, also add types to cacheName, enableSearch, label, and every other variable above

File: src/app/shared/components/fy-select/fy-select-modal/fy-select-modal.component.spec.ts [hunk 13]
```diff
@@ -126,8 +125,8 @@ describe('FySelectModalComponent', () => {
     component.recentrecentlyUsedItems$.pipe(take(1)).subscribe((res) => {
       expect(res).toEqual([{ label: 'business', value: 'BUSINESS' }]);
```
Comment (yksolanki9) [inline src/app/shared/components/fy-select/fy-select-modal/fy-select-modal.component.spec.ts:126]: Expect getRecentlyUsedItems to be called once
expect this.utilityService.searchArrayStream to be called once with searchTExt

File: src/app/shared/components/fy-select/fy-select-modal/fy-select-modal.component.spec.ts [hunk 14]
```diff
@@ -177,38 +176,34 @@ describe('FySelectModalComponent', () => {
...
-    component.currentSelection = 'BUSINESS';
+    it('should return recently used items from API if available', () => {
+      component.recentlyUsed = recentlyUsed;
+      component.options = [];
 
-    component.getRecentlyUsedItems().subscribe((result) => {
-      expect(getSpy).toHaveBeenCalledWith(component.cacheName);
-      expect(result).toEqual(filteredItems);
+      component.getRecentlyUsedItems().subscribe((result) => {
+        expect(result).toEqual(recentlyUsed);
```
Comment (yksolanki9) [inline src/app/shared/components/fy-select/fy-select-modal/fy-select-modal.component.spec.ts:190]: expect recentLocalStorageItemsService.get to not have been called

**Review by yksolanki9 — CHANGES_REQUESTED** [review]
Added comments

File: src/app/shared/components/fy-select/fy-select-modal/fy-select-modal.interface.ts [hunk 15]
```diff
@@ -0,0 +1,13 @@
+export interface Value {
```
Comment (yksolanki9) [inline src/app/shared/components/fy-select/fy-select-modal/fy-select-modal.interface.ts:1]: Rename to ModalOption or some relevant name.

**Review by yksolanki9 — APPROVED** [review]

**Review by Julias0 — APPROVED** [review]

**Pull request merged**